### PR TITLE
New feature: SOLD2 line detector and descriptor

### PR DESCRIPTION
**Pull request by rpautrat**
#### Changes
Addition of the SOLD2 line detector and descriptor. I separated the detector and descriptor to allow to only detect lines, or detect&describe and perform line matching across images.

The added code does not require any special package outside of numpy and pytorch.

Fixes https://github.com/kornia/kornia/issues/1124

#### Type of change

- [X] 📚  Documentation Update
- [X] 🧪 Tests Cases
- [X] 🔬 New feature (non-breaking change which adds functionality)
- [X] 📝 This change requires a documentation update

#### Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings

### CODE REVIEW

**Comment by ducha-aiki** [comment]
Wow, what a Christmas gift!

**Comment by rpautrat** [comment]
Yes, I finally found some time to do it!
Merry Christmas ;)

**Comment by edgarriba** [comment]
@rpautrat awesome and thanks! I'll review over the next week and target to be merged for the release after new year's :)

File: kornia/feature/sold2/sold2.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+import torch.nn.functional as F
+
+from .backbones import SOLD2Net
+from .sold2_detector import line_map_to_segments, LineSegmentDetectionModule, super_nms
+
+urls: Dict[str, str] = {}
+urls["wireframe"] = "https://www.polybox.ethz.ch/index.php/s/blOrW89gqSLoHOk/download"
+
+
+default_cfg = {
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:15]:
```suggestion
default_cfg: Dict = {
```

File: kornia/feature/sold2/sold2.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
+from typing import Dict
+
+import numpy as np
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:3]: no numpy -- we only support torch

File: kornia/feature/sold2/sold2.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        self.line_detector_cfg = self.config["line_detector_cfg"]
+        self.line_detector = LineSegmentDetectionModule(**self.config["line_detector_cfg"])
+
+        # Initialize the line matcher
+        self.line_matcher = WunschLineMatcher(**self.config["line_matcher_cfg"])
+
+    def forward(self, img: torch.Tensor) -> Dict:
+        """
+        Args:
+            img: batched images with shape :math:`(N, 1, H, W)`.
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:111]:
```suggestion
            img: batched images with shape :math:`(B, 1, H, W)`.
```

File: kornia/feature/sold2/sold2.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+        :return:
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:113]:
```suggestion
        Return:
```

File: kornia/feature/sold2/sold2.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+            - ``line_segments``: list of line segments in each of the N images :math:`List[(Nlines, 2, 2)]`.
+            - ``raw_junc_heatmap``: raw junction heatmap of shape :math:`(N, H, W)`.
+            - ``raw_line_heatmap``: raw line heatmap of shape :math:`(N, H, W)`.
+            - ``dense_desc``: the semi-dense descriptor map of shape :math:`(N, 128, H/4, W/4)`.
+
+        :rtype: dict
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:119]:
```suggestion
```
return type in principle is inferred

File: kornia/feature/sold2/sold2.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+        outputs["raw_line_heatmap"] = heatmap
+        heatmap_np = heatmap.detach().cpu().numpy()
+
+        # Loop through all images
+        lines = []
+        for i, curr_junc_pred_nms in enumerate(junc_pred_nms):
+            junctions = np.stack(np.where(curr_junc_pred_nms), axis=-1)
+            # Run the line detector
+            line_map, junctions, _ = self.line_detector.detect(
+                junctions, heatmap_np[i], device=device)
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:149]: can we implement this in pytorch ?

File: kornia/feature/sold2/sold2.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+        del state_dict["w_desc"]
+        state_dict["heatmap_decoder.conv_block_lst.2.0.weight"] = state_dict["heatmap_decoder.conv_block_lst.2.weight"]
+        state_dict["heatmap_decoder.conv_block_lst.2.0.bias"] = state_dict["heatmap_decoder.conv_block_lst.2.bias"]
+        del state_dict["heatmap_decoder.conv_block_lst.2.weight"]
+        del state_dict["heatmap_decoder.conv_block_lst.2.bias"]
+        return state_dict
+
+
+class WunschLineMatcher:
+    """Class matching two sets of line segments with the Needleman-Wunsch algorithm."""
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:184]: add a todo to move later in `kornia.feature.matching` /cc @ducha-aiki

File: kornia/feature/sold2/sold2.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+        for i in range(n):
+            for j in range(m):
+                nw_grid[:, i + 1, j + 1] = np.maximum(
+                    np.maximum(nw_grid[:, i + 1, j], nw_grid[:, i, j + 1]),
+                    nw_grid[:, i, j] + nw_scores[:, i, j])
+
+        return nw_grid[:, -1, -1]
+
+
+def keypoints_to_grid(keypoints: torch.Tensor, img_size: tuple) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:356]: we have https://kornia.readthedocs.io/en/latest/geometry.conversions.html#kornia.geometry.conversions.normalize_pixel_coordinates

File: kornia/feature/sold2/sold2.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+        self.line_matcher = WunschLineMatcher(**self.config["line_matcher_cfg"])
+
+    def forward(self, img: torch.Tensor) -> Dict:
+        """
+        Args:
+            img: batched images with shape :math:`(N, 1, H, W)`.
+
+        :return:
+            - ``line_segments``: list of line segments in each of the N images :math:`List[(Nlines, 2, 2)]`.
+            - ``raw_junc_heatmap``: raw junction heatmap of shape :math:`(N, H, W)`.
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2.py:115]: the convention we normally use is`B` for number of batches and in this case `N` could be number of lines ?

File: kornia/feature/sold2/sold2_detector.py [hunk 10]
```diff
@@ -0,0 +1,763 @@
...
+        state_dict["heatmap_decoder.conv_block_lst.2.0.weight"] = state_dict["heatmap_decoder.conv_block_lst.2.weight"]
+        state_dict["heatmap_decoder.conv_block_lst.2.0.bias"] = state_dict["heatmap_decoder.conv_block_lst.2.bias"]
+        del state_dict["heatmap_decoder.conv_block_lst.2.weight"]
+        del state_dict["heatmap_decoder.conv_block_lst.2.bias"]
+        return state_dict
+
+    def forward(self, img: torch.Tensor) -> Dict:
+        """
+        Args:
+            img: batched images with shape :math:`(N, 1, H, W)`.
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:104]: same comment as before for the convention

File: kornia/feature/sold2/sold2_detector.py [hunk 11]
```diff
@@ -0,0 +1,763 @@
...
+            raise ValueError("The input image should be a 4D torch tensor.")
+        device = img.device
+        outputs = {}
+
+        # Forward pass of the CNN backbone
+        net_outputs = self.model(img)
+
+        # Junction NMS
+        outputs["raw_junc_heatmap"] = net_outputs["junctions"]
+        junc_pred_nms = super_nms(
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:123]: implement this in pytorch

File: kornia/feature/sold2/sold2_detector.py [hunk 12]
```diff
@@ -0,0 +1,763 @@
...
+        self.inlier_thresh = inlier_thresh
+        self.local_patch_radius = max_local_patch_radius
+        self.lambda_radius = lambda_radius
+
+        # Detecting junctions on the boundary parameters
+        self.low_thresh = heatmap_low_thresh
+        self.high_thresh = heatmap_high_thresh
+
+        # Pre-compute the linspace sampler
+        self.sampler = np.linspace(0, 1, self.num_samples)
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:196]: pytorch

File: kornia/feature/sold2/sold2_detector.py [hunk 13]
```diff
@@ -0,0 +1,763 @@
...
+        else:
+            raise ValueError(
+                "[Error] Inputs must either be torch tensor or numpy ndarray.")
+
+        return outputs
+
+    def detect(self, junctions, heatmap, device=torch.device("cpu")):
+        """Main function performing line segment detection."""
+        # Convert inputs to torch tensor
+        junctions = self.convert_inputs(junctions, device=device)
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:230]: we should pass also `dtype` or promote it from `junctions` or `heatmap`

File: kornia/feature/sold2/sold2_detector.py [hunk 14]
```diff
@@ -0,0 +1,763 @@
...
+        # [Simple threshold detection]
+        # detection_results is a mask over all candidates
+        detection_results = (torch.mean(sampled_feat, dim=-1)
+                             > self.detect_thresh)
+
+        # [Inlier threshold detection]
+        if self.inlier_thresh > 0.:
+            inlier_ratio = torch.sum(
+                sampled_feat > self.detect_thresh,
+                dim=-1).to(torch.float32) / self.num_samples
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:341]:
```suggestion
                dim=-1).to(dtype) / self.num_samples
```

File: kornia/feature/sold2/sold2_detector.py [hunk 15]
```diff
@@ -0,0 +1,763 @@
...
+                      detected_junc_indexes[:, 0]] = 1
+
+        # [Junction refinement]
+        if self.use_junction_refinement and len(detected_junc_indexes) > 0:
+            junctions, line_map_pred = self.refine_junction_perturb(
+                junctions, line_map_pred, heatmap, H, W, device)
+
+        return line_map_pred, junctions, heatmap
+
+    def refine_heatmap(self, heatmap, ratio=0.2, valid_thresh=1e-2):
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:359]:
```suggestion
    def refine_heatmap(self, heatmap: Tensor, ratio: float = 0.2, valid_thresh: float = 1e-2) -> Tensor:
```
revisit if you have some time the typings

File: kornia/feature/sold2/sold2_detector.py [hunk 16]
```diff
@@ -0,0 +1,763 @@
...
+        sampled_feat = sampled_feat * patch_dist_mask.to(torch.float32)
+        if len(sampled_feat) == 0:
+            sampled_feat_lmax = torch.empty(0, 64)
+        else:
+            sampled_feat_lmax, _ = torch.max(sampled_feat, dim=-1)
+
+        return sampled_feat_lmax
+
+
+def line_map_to_segments(junctions: np.ndarray, line_map: np.ndarray) -> np.ndarray:
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:644]: implement in pytorch

File: kornia/feature/sold2/sold2_detector.py [hunk 17]
```diff
@@ -0,0 +1,763 @@
...
+                (output_segments, single_seg[None, ...]), axis=0)
+
+            # Update line_map
+            line_map_tmp[idx, idx2] = 0
+            line_map_tmp[idx2, idx] = 0
+
+    return output_segments
+
+
+def super_nms(prob_predictions: np.ndarray, dist_thresh: float,
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:670]: implement in pytorch -- also recently added https://kornia.readthedocs.io/en/latest/geometry.bbox.html?highlight=nms#kornia.geometry.bbox.nms

we could benchmark which it's best, or similarly I did with the face detector, make it easy to plug any (e.g torchvision has a nice c++/cuda compiled and quite fast) version that can be easily interfaced
https://github.com/kornia/kornia/blob/be104a5d783788bd5e37338d721e4caf506b0508/kornia/contrib/face_detection.py#L172

File: kornia/feature/sold2/sold2_detector.py [hunk 18]
```diff
@@ -0,0 +1,763 @@
...
+        # Re-compose the probability map
+        output_map = np.zeros([im_h, im_w])
+        output_map[keep_points[:, 0].astype(int),
+                   keep_points[:, 1].astype(int)] = keep_score.squeeze()
+        output_lst.append(output_map)
+
+    return np.stack(output_lst)
+
+
+def nms_fast(in_corners: np.ndarray, H: int, W: int, dist_thresh: float) -> np.ndarray:
```
Comment (edgarriba) [inline kornia/feature/sold2/sold2_detector.py:704]: which one we keep?

File: kornia/feature/sold2/backbones.py [hunk 19]
```diff
@@ -0,0 +1,436 @@
...
+    Args:
+        input_channel: number of input channels.
+        depth: number of residual blocks per hourglass module.
+        num_stacks: number of hourglass modules stacked together.
+        num_blocks: number of layers in each residual block.
+        num_classes: number of heads for the output of a hourglass module.
+    """
+
+    def __init__(self,
+                 input_channel: Optional[int] = 1,
```
Comment (edgarriba) [inline kornia/feature/sold2/backbones.py:25]:
```suggestion
                 input_channel: int = 1,
```
`Optional` is only used with `None`

File: kornia/feature/sold2/backbones.py [hunk 20]
```diff
@@ -0,0 +1,436 @@
...
+            out = self.pixshuffle(out)
+
+        # Output layer
+        out = self.conv_block_lst[-1](out)
+        heatmap = softmax(out, dim=1)[:, 1:, :, :]
+
+        return heatmap
+
+
+class SuperpointDescriptor(nn.Module):
```
Comment (edgarriba) [inline kornia/feature/sold2/backbones.py:366]: @ducha-aiki we should check later with other ongoing projects how we could reuse this

**Comment by rpautrat** [comment]
Thanks for the detailed review and feedback! I might need a bit of time for the re-implementation from numpy to pytorch because there are some deadlines coming soon.

The colab showing how to reproduce the GIF would be feasible, yes.

Regarding generalization and fine-tuning, it generalizes fairly well indoors, but the performance is not as good outdoors. I was planning to release an outdoor model at some point, but it takes some time to train and tune it properly. Fine-tuning is indeed cumbersome unfortunately, because you need to generate the pseudo ground truth on the new dataset each time... But if I manage to get a good outdoor model, it might be less essential to fine-tune it.

**Comment by edgarriba** [comment]
> Thanks for the detailed review and feedback! I might need a bit of time for the re-implementation from numpy to pytorch because there are some deadlines coming soon.
> 
> The colab showing how to reproduce the GIF would be feasible, yes.
> 
> Regarding generalization and fine-tuning, it generalizes fairly well indoors, but the performance is not as good outdoors. I was planning to release an outdoor model at some point, but it takes some time to train and tune it properly. Fine-tuning is indeed cumbersome unfortunately, because you need to generate the pseudo ground truth on the new dataset each time... But if I manage to get a good outdoor model, it might be less essential to fine-tune it.

Sounds good ! No rush here from our side. Let us know if you need help with any part.

File: kornia/feature/sold2/sold2_detector.py (hunk 17 above)
Comment (rpautrat) [reply kornia/feature/sold2/sold2_detector.py]: I did a small benchmark between the kornia version, the torchvision one, and the one I had in my code. The torchvision one is the fastest, both for CPU and GPU, so I am keeping this one.

**Comment by rpautrat** [comment]
Hi, I had time to work again on this PR.

I have moved all numpy functions to Pytorch as suggested, and opted for the torchvision nms operation. It was the fastest among the kornia one, torchvision one and SuperPoint one.

After consideration, releasing the code generating the GIF in a colab might not be possible, as it was partially based on the SuperPoint demo (https://github.com/magicleap/SuperPointPretrainedNetwork). The license stipulates that you can only use the code for research or non-profit purposes, so releasing it on Colab might be against it...
I can however do a Colab reimplementing the notebook that I had in my official SOLD2 release (https://github.com/cvg/SOLD2/blob/main/notebooks/match_lines.ipynb), but importing SOLD2 from kornia this time.

**Review by edgarriba — COMMENTED** [review]
@ducha-aiki @shijianjian can you review 

@rpautrat sounds good about the colab thing -- the example looks really nice

File: kornia/feature/sold2/sold2.py [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+        del state_dict["w_heatmap"]
+        del state_dict["w_desc"]
+        state_dict["heatmap_decoder.conv_block_lst.2.0.weight"] = state_dict["heatmap_decoder.conv_block_lst.2.weight"]
+        state_dict["heatmap_decoder.conv_block_lst.2.0.bias"] = state_dict["heatmap_decoder.conv_block_lst.2.bias"]
+        del state_dict["heatmap_decoder.conv_block_lst.2.weight"]
+        del state_dict["heatmap_decoder.conv_block_lst.2.bias"]
+        return state_dict
+
+
+class WunschLineMatcher:
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/feature/sold2/sold2.py:167]: Yes, I will take a look after this PR merged

**Review by ducha-aiki — APPROVED** [review]
Besides the @edgarriba comments about imports, LGTM

File: kornia/feature/sold2/sold2_detector.py [hunk 22]
```diff
@@ -0,0 +1,583 @@
+import math
+from typing import Dict, OrderedDict, Tuple
+
+import torch
+import torch.nn as nn
+from torchvision.ops import nms
```
[the remark this answers is not in the document]
Comment (rpautrat) [reply kornia/feature/sold2/sold2_detector.py]: I changed it to the kornia nms instead. It is slightly slower than the torchvision one, but there is no big difference in the end.

[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/feature/sold2/sold2_detector.py]: good then :)

kornia is purely tensor based -- while torchvision is specific kernel compiled (cpu/gpu). So, I expect that as pytorch gets mature we won't need at all then to have specific kernels, compile, etc and will be usable in many platforms

[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/feature/sold2/sold2_detector.py]: will play with it soon with Rust just to compare

**Comment by ducha-aiki** [comment]
@rpautrat I have added some minor fixes, preserving commit history here: https://github.com/kornia/kornia/pull/1844
Thank you a lot for contributing!
